### PR TITLE
Improve trunk eviction in EditManager

### PR DESCRIPTION
**Pull request by noencke**
## Description

The SharedTree EditManager's trunk commit eviction was both more complicated than necessary and also buggy. This PR cleans up the code and addresses the known bugs. Previously, when the minimum sequence number advanced to some sequence number _M_, all commits with sequence numbers below _M_ were dropped, but a commit with exactly sequence number _M_ was not because any number of branches might be based off of that commit (e.g. the local branch), and dropping it was impossible because a branch might continue to reference it. After this PR, that commit is still not dropped from memory, however it takes the place of the special "trunk base" commit or "origin" commit which precedes the trunk but is not considered part of the trunk. Not only does this remove multiple edge cases from the eviction logic (and the bugs that currently dwell therein), it also makes it more intuitive to predict which commits are removed from the trunk as the minimum sequence number advances.

image

This PR also does some minor cleanup and optimizations in EditManager, adds some more eviction test cases, and enables an eviction pass after tracked branches rebase (which will evict the trunk more aggressively in scenarios where a long-lived user-created branch repeatedly rebases over a parent branch).

### CODE REVIEW

File: experimental/dds/tree2/src/shared-tree-core/editManager.ts [hunk 1]
```diff
@@ -144,22 +152,17 @@ export class EditManager<
...
 		onForkTransitive(this.localBranch, (fork) => this.registerBranch(fork));
 	}
 
 	/**
 	 * Make the given branch known to the `EditManager`. The `EditManager` will ensure that all registered
 	 * branches remain usable even as the minimum sequence number advances.
-	 * @param branch - the branch to register. All branches that fork from this branch, directly or transitively,
-	 * will also be registered.
-	 */
+]	 */
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree-core/editManager.ts:165]: I'm guessing this was not intentional.

File: experimental/dds/tree2/src/shared-tree-core/editManager.ts [hunk 2]
```diff
@@ -243,59 +246,50 @@ export class EditManager<
...
+			// regenerating the entire commit graph. Instead, we can simply chop off the tail like this if we're certain
+			// that there are no outstanding references to any of the commits being removed (other than the references via
+			// the trunk). The peer branches have been rebased to the head of the trunk, the local branch is already rebased
+			// to the head of the trunk, and all other branches are tracked by `trunkBranches` and known to be ahead of or at
+			// `newTrunkBase`. Therefore, no branches should have unique references to any of the commits being evicted here.
+			// We mutate the most recent of the evicted commits to become the new trunk base. That way, any other commits that
+			// have parent pointers to the latest evicted commit will stay linked, even though that it is no longer part of the trunk.
+			const newTrunkBase = latestEvicted as Mutable<GraphCommit<TChangeset>>;
+			newTrunkBase.revision = this.trunkBase.revision;
+			newTrunkBase.change = this.trunkBase.change;
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree-core/editManager.ts:268]: Is it necessary for correctness to replace the revision and change of this commit? Or is it just for perf (because nothing should need those anymore)?

File: experimental/dds/tree2/src/shared-tree-core/editManager.ts [hunk 3]
```diff
@@ -243,59 +246,50 @@ export class EditManager<
...
+
+			// Update any state that is derived from trunk commits
+			this.sequenceMap.editRange(
+				minimumPossibleSequenceNumber,
+				sequenceNumber,
+				true,
+				(s, { revision }) => {
+					// Cleanup look-aside data for each evicted commit
+					this.trunkMetadata.delete(revision);
+					this.localBranchUndoRedoManager.untrackCommitType(revision);
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree-core/editManager.ts:280]: I realize this is not part of this PR, but I'm curious why we need to call `this.localBranchUndoRedoManager.untrackCommitType(revision);` here.

File: experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts [hunk 4]
```diff
@@ -260,40 +260,44 @@ describe("SharedTreeCore", () => {
...
-		factory.processAllMessages(); //          [x, x, x, 4, 5 (b5)]
+		factory.processAllMessages(); //                     [x, x, x (b4), 4, 5]
 		assert.equal(getTrunkLength(tree), 2);
-		branch5.dispose(); //                     [x, x, x, 4, 5]
+		const branch5 = tree.getLocalBranch().fork(); //     [x, x, x (b4), 4, 5 (b5)]
+		branch4.rebaseOnto(branch5); //                      [x, x, x, 4, 5 (b4, b5)]
+		branch4.dispose(); //                                [x, x, x, 4, 5 (b5)]
 		assert.equal(getTrunkLength(tree), 2);
+		changeTree(tree);
+		factory.processAllMessages(); //                     [x, x, x, x, 5 (b5)]
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts:296]: Shouldn't this be `[x, x, x, x, 5 (b5) 6]` because of the `changeTree(tree)` on line 295?

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts:296]: This will also affect line 299.

File: experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts [hunk 5]
```diff
@@ -260,40 +260,44 @@ describe("SharedTreeCore", () => {
...
+		assert.equal(getTrunkLength(tree), 1);
+		changeTree(tree);
+		branch5.dispose(); //                                [x, x, x, x, x, 6]
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts:299]: Not `[x, x, x, x, x, x, 7]`?

File: experimental/dds/tree2/src/shared-tree-core/editManager.ts (hunk 3 above)
Comment (noencke) [reply experimental/dds/tree2/src/shared-tree-core/editManager.ts:280]: @jenn-le, do you remember?

File: experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts (hunk 4 above)
Comment (noencke) [reply experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts:296]: Yes, you're right - thanks!

File: experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts (hunk 5 above)
Comment (noencke) [reply experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts:299]: Good catch.

File: experimental/dds/tree2/src/shared-tree-core/editManager.ts (hunk 3 above)
Comment (jenn-le) [reply experimental/dds/tree2/src/shared-tree-core/editManager.ts:280]: Yup, that's exactly it.

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/shared-tree-core/editManager.ts:280]: My mental model is that the UndoRedoManager's map from revision -> commit type should grow unbounded because we haven't yet added the feature to bound the undo/redo stack size.
If we are preventing that map from growing unbounded, then what happens when the user wants to undo a commit for which we've evicted the relevant data?

File: experimental/dds/tree2/src/shared-tree-core/editManager.ts (hunk 2 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/shared-tree-core/editManager.ts:268]: Makes sense. I think explaining as much in comments above lines 267 and 268 would be nice.

**Review by yann-achard-MS — APPROVED** [review]

File: experimental/dds/tree2/src/shared-tree-core/editManager.ts (hunk 2 above)
Comment (noencke) [reply experimental/dds/tree2/src/shared-tree-core/editManager.ts]: Good idea, I added some more explanatory comments.

File: experimental/dds/tree2/src/shared-tree-core/editManager.ts (hunk 3 above)
Comment (noencke) [reply experimental/dds/tree2/src/shared-tree-core/editManager.ts:284]: I'll investigate this offline.
